src/ocaml/driver/pparse.ml

```diff
@@ -158,8 +158,12 @@ let apply_pp ~workdir ~filename ~source ~pp =
     close_out oc
   end;
   let fn_out = fn_in ^ ".out" in
-  let comm = pp_commandline pp fn_in fn_out in
-  let ok = merlin_system_command comm ~cwd:workdir = 0 in
+  let args = [
+    Filename.quote fn_in;
+    Printf.sprintf "1>%s" (Filename.quote fn_out)
+  ] in
+  let comm = commandline pp args in
+  let ok = merlin_system_command ~prog:pp ~args:args ~cwd:workdir = 0 in
```

Comment:
I think this function should not expect its arguments to be shell-quoted.

Currently, merlin runs a shell command (`Sys.command "cd ... && ..."`) that does require specific quoting.

But users of the library that want to use something such as `Spawn` don't need  the arguments to be quoted since `Spawn` is directly using a C stub calling `execve` instead of relying on the shell.

Reply:
Current merlin fork used by ocaml-lsp works with quoting:
```ocaml
let merlin_system_command =
  if Sys.win32 then
    windows_merlin_system_command
  else
    fun cmd ~cwd ->
      let prog = "/bin/sh" in
      let argv = ["sh"; "-c"; cmd] in
      let stdin = Unix.openfile "/dev/null" [ Unix.O_RDONLY ] 0x777  in
      let pid =
        let cwd : Spawn.Working_dir.t = Path cwd in
        let stdout = Unix.stderr in
        Spawn.spawn ~cwd ~prog ~argv ~stdin ~stdout ()
      in
      let (_, status) = Unix.waitpid [] pid in
      let res =
        match (status : Unix.process_status) with
        | WEXITED n -> n
        | WSIGNALED _ -> -1
        | WSTOPPED _ -> -1
      in
      Unix.close stdin;
      res

let ppx_commandline cmd fn_in fn_out =
  Printf.sprintf "%s %s %s%s"
    cmd (Filename.quote fn_in) (Filename.quote fn_out)
    (if Sys.win32 then "" else " 1>&2")

let apply_rewriter magic ppx (fn_in, failures) =
  let title = "apply_rewriter" in
  let fn_out = Filename.temp_file "camlppx" "" in
  let comm = ppx_commandline ppx.workval fn_in fn_out in
  log ~title "running %s from directory %S" comm ppx.workdir;
  Logger.log_flush ();
  let failure =
    let ok = merlin_system_command comm ~cwd:ppx.workdir = 0 in
[...]
```

The problem is that we can't move filename quoting to `run_in_directory` function, because it takes list of arguments. We're not able to decide which are filenames nad needs to be quoted without changing it's signature.

Reply:
I wonder why is the shell still used with spawn. @rgrinberg ? 
Couldn't `[ppx_cmd; ppx_arg1; ...]` be used directly instead of `["sh"; "-c"; "quoted_ppx_cmd_with_args"]` ?

Reply:
I just never got around to changing it yet.